gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GHFSInstrumentation.java

```diff
@@ -0,0 +1,775 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.hadoop.fs.gcs;
+
+import static org.apache.hadoop.fs.statistics.IOStatisticsSupport.snapshotIOStatistics;
+import static org.apache.hadoop.fs.statistics.StoreStatisticNames.SUFFIX_FAILURES;
+import static org.apache.hadoop.fs.statistics.impl.IOStatisticsBinding.iostatisticsStore;
+
+import com.google.common.flogger.GoogleLogger;
+import java.io.Closeable;
+import java.net.URI;
+import java.time.Duration;
+import java.util.EnumSet;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nullable;
+import org.apache.hadoop.classification.InterfaceAudience;
+import org.apache.hadoop.classification.InterfaceStability;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.statistics.*;
+import org.apache.hadoop.fs.statistics.impl.IOStatisticsBinding;
+import org.apache.hadoop.fs.statistics.impl.IOStatisticsStore;
+import org.apache.hadoop.fs.statistics.impl.IOStatisticsStoreBuilder;
+import org.apache.hadoop.metrics2.MetricsCollector;
+import org.apache.hadoop.metrics2.MetricsSource;
+import org.apache.hadoop.metrics2.MetricsSystem;
+import org.apache.hadoop.metrics2.impl.MetricsSystemImpl;
+import org.apache.hadoop.metrics2.lib.MetricsRegistry;
+import org.apache.hadoop.metrics2.lib.MutableCounterLong;
+import org.apache.hadoop.metrics2.lib.MutableGaugeLong;
+import org.apache.hadoop.metrics2.lib.MutableMetric;
+
+/**
+ * Instrumentation of GCS.
+ *
+ * <p>History
+ *
+ * <ol>
+ *   <li>HADOOP-13028. Initial implementation. Derived from the {@code S3AInstrumentation}.
+ *   <li>Broadly (and directly) used in gcs. The use of direct references causes "problems" in
+ *       mocking tests.
+ *   <li>HADOOP-16830. IOStatistics. Move to an interface and implementation design for the
+ *       different inner classes.
+ * </ol>
+ *
+ * <p>Counters and metrics are generally addressed in code by their name or {@link GHFSStatistic}
+ * key. There <i>may</i> be some Statistics which do not have an entry here. To avoid attempts to
+ * access such counters failing, the operations to increment/query metric values are designed to
+ * handle lookup failures.
+ *
+ * <p>GoogleHadoopFileSystem StorageStatistics are dynamically derived from the IOStatistics.
+ */
+@InterfaceAudience.Private
+@InterfaceStability.Evolving
+class GHFSInstrumentation
+    implements Closeable, MetricsSource, IOStatisticsSource, DurationTrackerFactory {
+  public static final String CONTEXT = "DFSClient";
```

Comment:
What does it mean?

Reply:
@medb It is just to mention the metrics are of a DistributedFileSystem . We will change it to gcsFilesystem to make it more specific.